feat: create staked expenditure

joanna-pagepro opened the pull request:
## Description

Creating a Payment Builder action with the `Staking` decision method, when the Staked expenditure extension is enabled.

## Testing

* Step 1. Make sure that the Staked expenditure extension is installed and enabled and create a Payment builder action with the `Staking` decision method.
* Step 2. When a modal appears, click the `Create payment` button

## Diffs

**New stuff** ✨

* `CreateStakedExpenditureModal` component
* `FinalizeWithStakingInfo` component

**Changes** 🏗

* `ActionButtons` component - `onSubmitClick` prop added to pass the action to show a modal instead of submit
* `ActionSidebarContent`component - added `CreateStakedExpenditureModal` modal
* `DecisionMethodField` and `DecisionMethodRow` components - added `Staking` decision method
* `PaymentBuilderWidget` - if the expenditure is staked `FinalizeWithStakingInfo` is shown instead of `FinalizeWithPermissionsInfo` in `Create` and `Review` steps
* `Button` component - div wrapper added to the loader

**Deletions** ⚰️

* 

## TODO

Contributes to https://github.com/JoinColony/colonyCDapp/issues/2344

iamsamgibbs commented on src/redux/sagas/expenditures/createStakedExpenditure.ts:
```diff
@@ -239,6 +240,8 @@ function* createStakedExpenditure({
 
+    setTxHash?.(txHash);
+
     // @TODO: Remove during advanced payments UI wiring
     // eslint-disable-next-line no-console
     console.log('Created expenditure ID:', expenditureId.toString());
```
This can be removed now that createStakedExpenditure has been wired up

iamsamgibbs reviewed (approved):
Nice job, this looks good to me!

<img width="1058" alt="Screenshot 2024-05-10 at 08 46 51" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/1832d111-e732-4444-ba95-cdd229f1f042">

<img width="523" alt="Screenshot 2024-05-10 at 08 28 32" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/d0b60b93-3f7d-4772-9da1-bdbd52033218">

marcocolony reviewed (changes requested):
@joanna-pagepro Could you please have a look at these?

When trying to create a payment with a user that has insufficient locked tokens, the necessary validation is not showing up as per the [designs](https://www.figma.com/file/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?type=design&node-id=23853-196834&mode=design&t=VpKEEaoPkwkovddh-4)

Instead, the modal gets stuck on infinite loading and an error message is displayed in the Userhub.
![Export-1715702247339](https://github.com/JoinColony/colonyCDapp/assets/165801260/7c6793e6-cb2f-4230-8dde-4032b7920980)

![image](https://github.com/JoinColony/colonyCDapp/assets/165801260/88aeb225-f1aa-412f-9d83-3c98ed0351be)

Here's the error being thrown in the console: 
![image](https://github.com/JoinColony/colonyCDapp/assets/165801260/d177575b-01d2-48d7-a79f-ce316d6c51d1)

jakubcolony commented on src/components/v5/common/CompletedAction/partials/PaymentBuilder/partials/PaymentBuilderWidget/PaymentBuilderWidget.tsx:
```diff
@@ -91,7 +101,12 @@ const PaymentBuilderWidget: FC<PaymentBuilderWidgetProps> = ({ action }) => {
       heading: { label: formatText({ id: 'expenditure.createStage.label' }) },
-      content: (
+      content: isStaked ? (
+        <FinalizeWithStakingInfo
```
I believe this component (and related ones) originally displayed motion/expenditure finalization info but it's now used in more places such as creation in this case. Could the naming reflect that better as technically you cannot finalize with staking?

joanna-pagepro replied:
Will `StepFinalizedWithStakingInfo` be a better name? If not, maybe you can suggest something?

jakubcolony replied:
@joanna-pagepro sorry missed your comment, there's no finalization with staking, in this case it's the create step. `CreatedWithStakingInfo` or similar would make sense for this scenario (it's the only action you can do with staking). In other places where `FinalizeWithPermissionsInfo` might be a bit confusing (it's used in cancel, create, review and release steps), maybe something generic such as `ActionWithPermissionsInfo`?

jakubcolony reviewed (changes requested):
Great start on this PR @joanna-pagepro.

I noticed the validation error was not shown when I had no tokens activated. There is no horizontal padding on the "Required stake amount" container.
<img width="526" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/112586815/d2dcab71-bc3e-431a-9fa4-c37fc2b87f7b">

The idea behind staking is to allow users with no permissions to create a payment. Currently I'm unable to do that:
<img width="698" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/112586815/f2b3bb6d-293d-406e-b09c-fa1bd0255c3f">


The create info card itself looks good:
<img width="391" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/112586815/a8c834b9-5a73-4595-9f47-b1d2b28b9607">

My stake was reclaimed after finalizing the payment.

marcocolony reviewed (changes requested):
Hey @joanna-pagepro, thanks for looking at this again. 

I've tried to go through the flow with a member with no permissions. I'm able to create the payment, but when I get to the funding step I'm able to select the permissions decision method and click on fund. Nothing happens after and I'm stuck on this step without an error message explaining why  


![Export-1716457088795](https://github.com/JoinColony/colonyCDapp/assets/165801260/f5d86c94-e8be-46d9-bf08-c47e7a97e442)

<img width="1800" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/165801260/15dab284-8fda-4980-b671-ee871e218c5c">

jakubcolony reviewed (approved):
Code-wise changes look good, nice work @joanna-pagepro 👍

arrenv commented:
Thank you @joanna-pagepro  and @marcocolony for raising this one.

- Any decision methods used throughout the process should follow the same rules as the action panel decision methods. i.e.
    - Only available ones should show.
    - If no other ones are installed, but the user does not have permissions, then validation should show that they don’t have permissions. With the select disabled. [Figma design link.](https://www.figma.com/design/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?node-id=24483-442557&t=q1re7UWcQs2Bzrf2-4)
    - If reputation is installed, the there is no reputation in the team, then validation should show that there is no reputation to create it. Shown when selecting the Reputation decision method. [Figma design link.](https://www.figma.com/design/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?node-id=24485-442966&t=q1re7UWcQs2Bzrf2-4)

joanna-pagepro commented:
@marcocolony can you review it again?

marcocolony reviewed (changes requested):
Hey @joanna-pagepro thanks for tackling this one again.

I'm still getting the same behaviour, although the console now displays an error stating that the user does not hold the necessary permissions.

<img width="1794" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/165801260/86f324d8-0fe9-4c5c-a384-78135a3d0a6a">


![Export-1716801206190](https://github.com/JoinColony/colonyCDapp/assets/165801260/ffaf1e4a-e5e3-4685-b9e3-1713b1a64252)

<img width="796" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/165801260/9581e5af-c22d-4ff1-9919-230c715a69f2">

joanna-pagepro commented:
@marcocolony can you check again? I'm not getting this error. Decision method select is disabled and you can see a box with the info about insufficient permissions.

@jakubcolony naming updated

marcocolony reviewed (changes requested):
@joanna-pagepro 

Tested again, the validation is appearing correctly now 🙌 

I am however getting the UI bug where the network badge, user hub and kebab icon are sticking out when the funding modal is active. I think I also reported this on another PR recently? 

![Export-1716911330886](https://github.com/JoinColony/colonyCDapp/assets/165801260/038a5118-d4b5-41ad-b88b-ee9095bd04bd)
<img width="1798" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/165801260/7f88b4fd-3a82-4934-a25a-807aea23885f">

arrenv reviewed (changes requested):
@joanna-pagepro Thank you for putting this one together. Very nice!

1. Reputation was not loading in local dev and it was showing I could create a payment with 0 required tokens. I was able to click the button, but it was stuck in a loading state with the following error in console.

   Can we handle this case by showing an error message in the same red callout card with the text "Reputation in the colony could not loaded, please reload the page and try again." and have the button reload the error on click, instead of trying to trigger the transaction.

![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/46e7bd02-7bfb-45b7-b396-bbd061cafdbe)


> I am however getting the UI bug where the network badge, user hub and kebab icon are sticking out when the funding modal is active. I think I also reported this on another PR recently?

@marcocolony this is a feature rather then a bug, when you need to perform a transaction, you need to be able to access these menu items. It is a little bit of a quirk, I think if anything the dark background behind the modal could be removed to avoid the oddness of it.

joanna-pagepro commented:
@arrenv @marcocolony action updated, can you review again?

marcocolony reviewed (approved):
@joanna-pagepro Nice work, I can see the warning message when reputation isn't loaded in the colony.

<img width="615" alt="image" src="https://github.com/JoinColony/colonyCDapp/assets/165801260/4e2c391b-92df-486f-8773-f3d8dcccc64a">

arrenv reviewed (changes requested):
@joanna-pagepro Thank you for the updates.

1. We need the same treatment here for the header to be visible, so that we can interact with the transaction.
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/96fab941-71b5-4b3c-b5b6-1ad2eb76443e)


2. I think there has been some confusion with the activated tokens and validation here, please check this from the spec:
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/79a03928-d844-4d3c-bb80-ba4520c457ca)

      Staking should automatically handle activating tokens, as it does during the motion process, the not enough funds validation is relevant only if they don't have enough funds taking into account their locked tokens, not already staked, and any unlocked funds.

![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/b48e9f7c-2a55-4f44-be9a-9341a76dacfc)

![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/90df5dcb-125e-48e8-8383-0249c296811c)


3. Not sure if introduced in this issue or not, but I don't see it in production, after creating, I can see both Review and Passed status pills in the header. When funding I see passed.
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/fea8e785-ccbe-4698-9b1c-fadaf0c35e15)
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/9f7a5b8e-61d0-4abe-9670-54fd558e617d)

4. The permission decision method is still visible, despite the users not having permissions, it is not selectable, but it should also not be visible as there is no feedback as to why they cannot select it.
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/86e54917-2f29-41ba-8948-a335a678ff8e)

joanna-pagepro commented:
@arrenv 
fixes for 1 and 3 should be already merged to the master or feat/advanced-payments-ui branch
2. This saga throws an error if you don't have enough tokens activated and that's why I implemented it like this. We may need some more changes here. @jakubcolony can you help here?
4. updated

joanna-pagepro commented:
@arrenv 2 should be also ready to review now

joanna-pagepro commented:
@arrenv can you check this one? :)

arrenv reviewed (changes requested):
@joanna-pagepro Thank you for the update and details.

All else looks good, but I am seeing this issue when funding the payment.

![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/9165e1ba-a1cb-4190-b160-f49d3b8ce95c)

joanna-pagepro commented:
@arrenv should be ok now. Can you check?

arrenv reviewed (approved):
@joanna-pagepro Everything looks good now, thank you for your hard work on this!
